lib/IRGen/GenBuiltin.cpp

```diff
@@ -1543,3 +1545,371 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
+    // We always pad out the stride, even for the final element.
+    return Size(arraySize * elementTI.getFixedStride().getValue());
+  }
+  
+  static SpareBitVector getArraySpareBits(unsigned arraySize,
+                                          const LoadableTypeInfo &elementTI) {
+    if (arraySize == 0) {
+      return SpareBitVector();
+    }
+    
+    // Take spare bits from the first element only.
+    SpareBitVector result = elementTI.getSpareBits();
+    result.appendClearBits(getArraySize(arraySize, elementTI).getValueInBits()
+                             - result.size());
+    return result;
+  }
+  
+  const unsigned ArraySize;
+  const LoadableTypeInfo &Element;
+  
+  static SILType getElementSILType(IRGenModule &IGM,
+                                   SILType arrayType) {
+    return IGM.getLoweredType(AbstractionPattern::getOpaque(),
+                  arrayType.castTo<BuiltinFixedArrayType>()->getElementType());
+  }
+  
+public:
+  LoadableArrayTypeInfo(unsigned arraySize,
+                        const LoadableTypeInfo &elementTI)
+    : LoadableTypeInfo(getArrayType(arraySize, elementTI),
+                       getArraySize(arraySize, elementTI),
+                       getArraySpareBits(arraySize, elementTI),
+                       elementTI.getFixedAlignment(),
+                       elementTI.isTriviallyDestroyable(ResilienceExpansion::Maximal),
+                       elementTI.isCopyable(ResilienceExpansion::Maximal),
+                       elementTI.isFixedSize(ResilienceExpansion::Minimal),
+                       elementTI.isABIAccessible()),
+      ArraySize(arraySize), Element(elementTI)
+  {
+  }
+  
+  unsigned getExplosionSize() const override {
+    return Element.getExplosionSize() * ArraySize;
+  }
+
+  template<typename Body>
+  void eachElementAddr(IRGenFunction &IGF,
+                       Address addr,
+                       Body &&body) const {
+    for (unsigned i = 0; i < ArraySize; ++i) {
+      auto elementAddr = IGF.Builder.CreateArrayGEP(addr,
+        llvm::ConstantInt::get(IGF.IGM.IntPtrTy, i), Element.getFixedStride());
+      
+      body(elementAddr);
+    }
+  }
+  template<typename Body>
+  void eachElement(Body &&body) const {
+    for (unsigned i = 0; i < ArraySize; ++i) {
+      body();
+    }
+  }
+
+  void getSchema(ExplosionSchema &schema) const override {
+    eachElement([&]{
+      Element.getSchema(schema);
+    });
+  }
+  
+  TypeLayoutEntry *
+  buildTypeLayoutEntry(IRGenModule &IGM,
+                       SILType T,
+                       bool useStructLayouts) const override {
+    std::vector<TypeLayoutEntry *> elements;
+    auto eltTy = getElementSILType(IGM, T);
+    eachElement([&]{
+      elements.push_back(Element.buildTypeLayoutEntry(IGM, eltTy,
+                                                      useStructLayouts));
+    });
+    return IGM.typeLayoutCache.getOrCreateAlignedGroupEntry(elements, T,
```

Comment:
You should be able to use the new `ArrayLayoutEntry` which takes an element layout, element type, and a count type (which is trivial to build back from an unsigned `IntegerType::get(std::to_string(arraySize), false, ctx)`).